File: runtime/Cpp/runtime/Lexer.cpp
#include "Lexer.h"

#include "LexerATNSimulator.h"
#include "Exceptions.h"
#include "ANTLRErrorListener.h"
#include "Interval.h"
#include "StringBuilder.h"
#include "CommonTokenFactory.h"
#include "LexerNoViableAltException.h"

/*
 * [The "BSD license"]
 *  Copyright (c) 2016 Mike Lischke
 *  Copyright (c) 2013 Terence Parr
 *  Copyright (c) 2013 Dan McLaughlin
 *  All rights reserved.
 *
 *  Redistribution and use in source and binary forms, with or without
 *  modification, are permitted provided that the following conditions
 *  are met:
 *
 *  1. Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *  2. Redistributions in binary form must reproduce the above copyright
 *     notice, this list of conditions and the following disclaimer in the
 *     documentation and/or other materials provided with the distribution.
 *  3. The name of the author may not be used to endorse or promote products
 *     derived from this software without specific prior written permission.
 *
 *  THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 *  IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 *  OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 *  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 *  NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 *  DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 *  THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 *  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 *  THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

namespace org {
    namespace antlr {
        namespace v4 {
            namespace runtime {


                Lexer::Lexer() {
                    InitializeInstanceFields();
                }

                Lexer::Lexer(CharStream *input) {
                    InitializeInstanceFields();
                    this->_input = input;
                    this->_tokenFactorySourcePair = new std::pair<TokenSource*, CharStream*>(this, input);
                }

                void Lexer::reset() {
                    // wack Lexer state variables
                    if (_input != nullptr) {
                        _input->seek(0); // rewind the input
                    }
                    delete _token;
                    _type = Token::INVALID_TYPE;
                    _channel = Token::DEFAULT_CHANNEL;
                    _tokenStartCharIndex = -1;
                    _tokenStartCharPositionInLine = -1;
                    _tokenStartLine = -1;
                    _text = L"";

                    _hitEOF = false;
                    _mode = Lexer::DEFAULT_MODE;
                    _modeStack.clear();

                    getInterpreter()->reset();
                }

                Token *Lexer::nextToken() {
                    if (_input == nullptr) {
                        throw new IllegalStateException(L"nextToken requires a non-null input stream.");
                    }

                    // Mark start location in char stream so unbuffered streams are
                    // guaranteed at least have text of current token
                    int tokenStartMarker = _input->mark();
                    try {
                        while (true) {
                            outerContinue:
                            if (_hitEOF) {
                                emitEOF();
                                return _token;
                            }

                            delete _token;
                            _channel = Token::DEFAULT_CHANNEL;
                            _tokenStartCharIndex = _input->index();
                            _tokenStartCharPositionInLine = getInterpreter()->getCharPositionInLine();
                            _tokenStartLine = getInterpreter()->getLine();
                            _text = L"";
                            do {
                                _type = Token::INVALID_TYPE;
                                //				System.out.println("nextToken line "+tokenStartLine+" at "+((char)input.LA(1))+
                                //								   " in mode "+mode+
                                //								   " at index "+input.index());
                                int ttype;
                                try {
                                    ttype = getInterpreter()->match(_input, _mode);
                                } catch (LexerNoViableAltException *e) {
                                    notifyListeners(e); // report error
                                    recover(e);
                                    ttype = SKIP;
                                }
                                if (_input->LA(1) == IntStream::_EOF) {
                                    _hitEOF = true;
                                }
                                if (_type == Token::INVALID_TYPE) {
                                    _type = ttype;
                                }
                                if (_type == SKIP) {
                                    goto outerContinue;
                                }
                            } while (_type == MORE);
                            if (_token == nullptr) {
                                emit();
                            }
                            return _token;
                        }
                        
                    }
                    catch(...) {
#ifdef TODO
                        // Do something intelligent here for once
#endif
                    }
                    
                    // make sure we release marker after match or
                    // unbuffered char stream will keep buffering
                    _input->release(tokenStartMarker);
                    return nullptr;
                }

                void Lexer::skip() {
                    _type = SKIP;
                }

                void Lexer::more() {
                    _type = MORE;
                }

                void Lexer::mode(int m) {
                    _mode = m;
                }

                void Lexer::pushMode(int m) {
                    if (atn::LexerATNSimulator::debug) {
                        std::wcout << std::wstring(L"pushMode ") << m << std::endl;
                    }
                    _modeStack.push_back(_mode);
                    mode(m);
                }

                int Lexer::popMode() {
                    if (_modeStack.empty()) {
                        throw EmptyStackException();
                    }
                    if (atn::LexerATNSimulator::debug) {
                        std::wcout << std::wstring(L"popMode back to ") << _modeStack.back() << std::endl;
                    }
                    mode(_modeStack.back());
                    _modeStack.pop_back();
                    return _mode;
                }
                

                TokenFactory<CommonToken*> *Lexer::getTokenFactory() {
                    return _factory;
                }

                void Lexer::setInputStream(IntStream *input) {
                    delete this->_input;
                    this->_tokenFactorySourcePair = new std::pair<TokenSource*, CharStream*>(this, _input);
                    reset();
                    this->_input = static_cast<CharStream*>(input);
                    this->_tokenFactorySourcePair = new std::pair<TokenSource*, CharStream*>(this, _input);
                }

                std::string Lexer::getSourceName() {
                    return _input->getSourceName();
                }

                CharStream *Lexer::getInputStream() {
                    return _input;
                }

                void Lexer::emit(Token *token) {
                    //System.err.println("emit "+token);
                    this->_token = token;
                }

                Token *Lexer::emit() {
                    Token *t = (Token*)_factory->create(_tokenFactorySourcePair, _type, _text, _channel, _tokenStartCharIndex, getCharIndex() - 1, _tokenStartLine, _tokenStartCharPositionInLine);
                    emit(t);
                    return t;
                }

                Token *Lexer::emitEOF() {
                    int cpos = getCharPositionInLine();
                    // The character position for EOF is one beyond the position of
                    // the previous token's last character
                    if (_token != nullptr) {
                        int n = _token->getStopIndex() - _token->getStartIndex() + 1;
                        cpos = _token->getCharPositionInLine() + n;
                    }
                    Token *eof = (Token*)_factory->create(_tokenFactorySourcePair, Token::_EOF, L"", Token::DEFAULT_CHANNEL, _input->index(), _input->index() - 1, getLine(), cpos);
                    emit(eof);
                    return eof;
                }

                int Lexer::getLine() {
                    return getInterpreter()->getLine();
                }

                int Lexer::getCharPositionInLine() {
                    return getInterpreter()->getCharPositionInLine();
                }

                void Lexer::setLine(int line) {
                    getInterpreter()->setLine(line);
                }

                void Lexer::setCharPositionInLine(int charPositionInLine) {
                    getInterpreter()->setCharPositionInLine(charPositionInLine);
                }

                int Lexer::getCharIndex() {
                    return _input->index();
                }

                std::wstring Lexer::getText() {
                    if (_text != L"") {
                        return _text;
                    }
                    return getInterpreter()->getText(_input);
                }

                void Lexer::setText(const std::wstring &text) {
                    this->_text = text;
                }

                Token *Lexer::getToken() {
                    return _token;
                }

                void Lexer::setToken(Token *_token) {
                    this->_token = _token;
                }

                void Lexer::setType(int ttype) {
                    _type = ttype;
                }

                int Lexer::getType() {
                    return _type;
                }

                void Lexer::setChannel(int channel) {
                    _channel = channel;
                }

                int Lexer::getChannel() {
                    return _channel;
                }

                std::vector<Token*> Lexer::getAllTokens() {
                    std::vector<Token*> tokens = std::vector<Token*>();
                    Token *t = nextToken();
                    while (t->getType() != Token::_EOF) {
                        tokens.push_back(t);
                        t = nextToken();
                    }
                    return tokens;
                }

                void Lexer::recover(LexerNoViableAltException *e) {
                    if (_input->LA(1) != IntStream::_EOF) {
                        // skip a char and try again
                        getInterpreter()->consume(_input);
                    }
                }

                void Lexer::notifyListeners(LexerNoViableAltException *e) {
                    std::wstring text = _input->getText(misc::Interval::of(_tokenStartCharIndex, _input->index()));
                    std::wstring msg = std::wstring(L"token recognition error at: '") + getErrorDisplay(text) + std::wstring(L"'");

                    ANTLRErrorListener *listener = getErrorListenerDispatch();
                    listener->syntaxError(this, nullptr, _tokenStartLine, _tokenStartCharPositionInLine, msg, e);
                }

                std::wstring Lexer::getErrorDisplay(const std::wstring &s) {
                    antlrcpp::StringBuilder *buf = new antlrcpp::StringBuilder();
                    
                    for (size_t i = 0; i < s.length(); i++) {
                        char c = ((char*)s.c_str())[i];
                        buf->append(getErrorDisplay(c));
                    }
                    /*
                    for (auto c : s.toCharArray()) {
                        buf->append(getErrorDisplay(c));
                    }*/

                    return buf->toString();
                }

                std::wstring Lexer::getErrorDisplay(int c) {
                    std::wstring s = antlrcpp::StringConverterHelper::toString(static_cast<wchar_t>(c));
                    switch (c) {
                        case Token::_EOF :
                            s = L"<EOF>";
                            break;
                        case L'\n' :
                            s = L"\\n";
                            break;
                        case L'\t' :
                            s = L"\\t";
                            break;
                        case L'\r' :
                            s = L"\\r";
                            break;
                    }
                    return s;
                }

                std::wstring Lexer::getCharErrorDisplay(int c) {
                    std::wstring s = getErrorDisplay(c);
                    return std::wstring(L"'") + s + std::wstring(L"'");
                }

                void Lexer::recover(RecognitionException *re) {
                    //System.out.println("consuming char "+(char)input.LA(1)+" during recovery");
                    //re.printStackTrace();
                    // TODO: Do we lose character or line position information?
                    _input->consume();
                }

                void Lexer::InitializeInstanceFields() {
                    _factory = CommonTokenFactory::DEFAULT;
                    _tokenStartCharIndex = -1;
                    _tokenStartLine = 0;
                    _tokenStartCharPositionInLine = 0;
                    _hitEOF = false;
                    _channel = 0;
                    _type = 0;
                    _mode = Lexer::DEFAULT_MODE;
                }
            }
        }
    }
}

File: runtime/Go/antlr/lexer.go
package antlr

import (
	"fmt"
	"strconv"
)

// A lexer is recognizer that draws input symbols from a character stream.
//  lexer grammars result in a subclass of this object. A Lexer object
//  uses simplified Match() and error recovery mechanisms in the interest
//  of speed.
///

type Lexer interface {
	TokenSource
	Recognizer

	setChannel(int)
	pushMode(int)
	popMode() int
	setType(int)
	mode(int)
}

type BaseLexer struct {
	*BaseRecognizer

	Interpreter         *LexerATNSimulator
	TokenStartCharIndex int
	TokenStartLine      int
	TokenStartColumn    int
	ActionType          int

	_input                  CharStream
	_factory                TokenFactory
	_tokenFactorySourcePair *TokenSourceCharStreamPair
	_token                  Token
	_hitEOF                 bool
	_channel                int
	_type                   int
	_modeStack              IntStack
	_mode                   int
	_text                   string
}

func NewBaseLexer(input CharStream) *BaseLexer {

	lexer := new(BaseLexer)

	lexer.BaseRecognizer = NewBaseRecognizer()

	lexer._input = input
	lexer._factory = CommonTokenFactoryDEFAULT
	lexer._tokenFactorySourcePair = &TokenSourceCharStreamPair{lexer, input}

	lexer.Interpreter = nil // child classes must populate it

	// The goal of all lexer rules/methods is to create a token object.
	// l is an instance variable as multiple rules may collaborate to
	// create a single token. NextToken will return l object after
	// Matching lexer rule(s). If you subclass to allow multiple token
	// emissions, then set l to the last token to be Matched or
	// something nonnil so that the auto token emit mechanism will not
	// emit another token.
	lexer._token = nil

	// What character index in the stream did the current token start at?
	// Needed, for example, to get the text for current token. Set at
	// the start of NextToken.
	lexer.TokenStartCharIndex = -1

	// The line on which the first character of the token resides///
	lexer.TokenStartLine = -1

	// The character position of first character within the line///
	lexer.TokenStartColumn = -1

	// Once we see EOF on char stream, next token will be EOF.
	// If you have DONE : EOF  then you see DONE EOF.
	lexer._hitEOF = false

	// The channel number for the current token///
	lexer._channel = TokenDefaultChannel

	// The token type for the current token///
	lexer._type = TokenInvalidType

	lexer._modeStack = make([]int, 0)
	lexer._mode = LexerDefaultMode

	// You can set the text for the current token to override what is in
	// the input char buffer. Use setText() or can set l instance var.
	// /
	lexer._text = ""

	return lexer
}

const (
	LexerDefaultMode = 0
	LexerMore        = -2
	LexerSkip        = -3
)

const (
	LexerDefaultTokenChannel = TokenDefaultChannel
	LexerHidden              = TokenHiddenChannel
	LexerMinCharValue        = '\u0000'
	LexerMaxCharValue        = '\uFFFE'
)

func (b *BaseLexer) reset() {
	// wack Lexer state variables
	if b._input != nil {
		b._input.Seek(0) // rewind the input
	}
	b._token = nil
	b._type = TokenInvalidType
	b._channel = TokenDefaultChannel
	b.TokenStartCharIndex = -1
	b.TokenStartColumn = -1
	b.TokenStartLine = -1
	b._text = ""

	b._hitEOF = false
	b._mode = LexerDefaultMode
	b._modeStack = make([]int, 0)

	b.Interpreter.reset()
}

func (b *BaseLexer) GetInterpreter() *LexerATNSimulator {
	return b.Interpreter
}

func (b *BaseLexer) GetInputStream() CharStream {
	return b._input
}

func (b *BaseLexer) GetSourceName() string {
	return b.GrammarFileName
}

func (b *BaseLexer) setChannel(v int) {
	b._channel = v
}

func (b *BaseLexer) GetTokenFactory() TokenFactory {
	return b._factory
}

func (b *BaseLexer) setTokenFactory(f TokenFactory) {
	b._factory = f
}

func (b *BaseLexer) safeMatch() (ret int) {

	// previously in catch block
	defer func() {
		if e := recover(); e != nil {
			if re, ok := e.(RecognitionException); ok {
				b.notifyListeners(re) // Report error
				b.Recover(re)
				ret = LexerSkip // default
			}
		}
	}()

	return b.Interpreter.Match(b._input, b._mode)
}

// Return a token from l source i.e., Match a token on the char stream.
func (b *BaseLexer) NextToken() Token {
	if b._input == nil {
		panic("NextToken requires a non-nil input stream.")
	}

	var tokenStartMarker = b._input.Mark()

	// previously in finally block
	defer func() {
		// make sure we release marker after Match or
		// unbuffered char stream will keep buffering
		b._input.Release(tokenStartMarker)
	}()

	for true {
		if b._hitEOF {
			b.emitEOF()
			return b._token
		}
		b._token = nil
		b._channel = TokenDefaultChannel
		b.TokenStartCharIndex = b._input.Index()
		b.TokenStartColumn = b.Interpreter.column
		b.TokenStartLine = b.Interpreter.line
		b._text = ""
		var continueOuter = false
		for true {
			b._type = TokenInvalidType
			var ttype = LexerSkip

			ttype = b.safeMatch()

			if b._input.LA(1) == TokenEOF {
				b._hitEOF = true
			}
			if b._type == TokenInvalidType {
				b._type = ttype
			}
			if b._type == LexerSkip {
				continueOuter = true
				break
			}
			if b._type != LexerMore {
				break
			}
			if PortDebug {
				fmt.Println("lex inner loop")
			}
		}

		if PortDebug {
			fmt.Println("lex loop")
		}
		if continueOuter {
			continue
		}
		if b._token == nil {
			b.emit()
		}
		return b._token
	}

	return nil
}

// Instruct the lexer to Skip creating a token for current lexer rule
// and look for another token. NextToken() knows to keep looking when
// a lexer rule finishes with token set to SKIP_TOKEN. Recall that
// if token==nil at end of any token rule, it creates one for you
// and emits it.
// /
func (b *BaseLexer) Skip() {
	b._type = LexerSkip
}

func (b *BaseLexer) More() {
	b._type = LexerMore
}

func (b *BaseLexer) mode(m int) {
	b._mode = m
}

func (b *BaseLexer) pushMode(m int) {
	if LexerATNSimulatorDebug {
		fmt.Println("pushMode " + strconv.Itoa(m))
	}
	b._modeStack.Push(b._mode)
	b.mode(m)
}

func (b *BaseLexer) popMode() int {
	if len(b._modeStack) == 0 {
		panic("Empty Stack")
	}
	if LexerATNSimulatorDebug {
		fmt.Println("popMode back to " + fmt.Sprint(b._modeStack[0:len(b._modeStack)-1]))
	}
	i, _ := b._modeStack.Pop()
	b.mode(i)
	return b._mode
}

func (b *BaseLexer) inputStream() CharStream {
	return b._input
}

func (b *BaseLexer) setInputStream(input CharStream) {
	b._input = nil
	b._tokenFactorySourcePair = &TokenSourceCharStreamPair{b, b._input}
	b.reset()
	b._input = input
	b._tokenFactorySourcePair = &TokenSourceCharStreamPair{b, b._input}
}

// By default does not support multiple emits per NextToken invocation
// for efficiency reasons. Subclass and override l method, NextToken,
// and GetToken (to push tokens into a list and pull from that list
// rather than a single variable as l implementation does).
// /
func (b *BaseLexer) emitToken(token Token) {
	b._token = token
}

// The standard method called to automatically emit a token at the
// outermost lexical rule. The token object should point into the
// char buffer start..stop. If there is a text override in 'text',
// use that to set the token's text. Override l method to emit
// custom Token objects or provide a Newfactory.
// /
func (b *BaseLexer) emit() Token {
	if PortDebug {
		fmt.Println("emit")
	}
	var t = b._factory.Create(b._tokenFactorySourcePair, b._type, b._text, b._channel, b.TokenStartCharIndex, b.getCharIndex()-1, b.TokenStartLine, b.TokenStartColumn)
	b.emitToken(t)
	return t
}

func (b *BaseLexer) emitEOF() Token {
	cpos := b.GetCharPositionInLine()
	lpos := b.GetLine()
	if PortDebug {
		fmt.Println("emitEOF")
	}
	var eof = b._factory.Create(b._tokenFactorySourcePair, TokenEOF, "", TokenDefaultChannel, b._input.Index(), b._input.Index()-1, lpos, cpos)
	b.emitToken(eof)
	return eof
}

func (b *BaseLexer) GetCharPositionInLine() int {
	return b.Interpreter.column
}

func (b *BaseLexer) GetLine() int {
	return b.Interpreter.line
}

func (b *BaseLexer) getType() int {
	return b._type
}

func (b *BaseLexer) setType(t int) {
	b._type = t
}

// What is the index of the current character of lookahead?///
func (b *BaseLexer) getCharIndex() int {
	return b._input.Index()
}

// Return the text Matched so far for the current token or any text override.
//Set the complete text of l token it wipes any previous changes to the text.
func (b *BaseLexer) GetText() string {
	if b._text != "" {
		return b._text
	}

	return b.Interpreter.GetText(b._input)
}

func (b *BaseLexer) SetText(text string) {
	b._text = text
}

func (b *BaseLexer) GetATN() *ATN {
	return b.Interpreter.atn
}

// Return a list of all Token objects in input char stream.
// Forces load of all tokens. Does not include EOF token.
// /
func (b *BaseLexer) getAllTokens() []Token {
	if PortDebug {
		fmt.Println("getAllTokens")
	}
	var tokens = make([]Token, 0)
	var t = b.NextToken()
	for t.GetTokenType() != TokenEOF {
		tokens = append(tokens, t)
		if PortDebug {
			fmt.Println("getAllTokens")
		}
		t = b.NextToken()
	}
	return tokens
}

func (b *BaseLexer) notifyListeners(e RecognitionException) {
	var start = b.TokenStartCharIndex
	var stop = b._input.Index()
	var text = b._input.GetTextFromInterval(NewInterval(start, stop))
	var msg = "token recognition error at: '" + text + "'"
	var listener = b.GetErrorListenerDispatch()
	listener.SyntaxError(b, nil, b.TokenStartLine, b.TokenStartColumn, msg, e)
}

func (b *BaseLexer) getErrorDisplayForChar(c rune) string {
	if c == TokenEOF {
		return "<EOF>"
	} else if c == '\n' {
		return "\\n"
	} else if c == '\t' {
		return "\\t"
	} else if c == '\r' {
		return "\\r"
	} else {
		return string(c)
	}
}

func (b *BaseLexer) getCharErrorDisplay(c rune) string {
	return "'" + b.getErrorDisplayForChar(c) + "'"
}

// Lexers can normally Match any char in it's vocabulary after Matching
// a token, so do the easy thing and just kill a character and hope
// it all works out. You can instead use the rule invocation stack
// to do sophisticated error recovery if you are in a fragment rule.
// /
func (b *BaseLexer) Recover(re RecognitionException) {
	if b._input.LA(1) != TokenEOF {
		if _, ok := re.(*LexerNoViableAltException); ok {
			// Skip a char and try again
			b.Interpreter.consume(b._input)
		} else {
			// TODO: Do we lose character or line position information?
			b._input.Consume()
		}
	}
}
